core/src/main/scala/com/github/tminglei/slickpg/ExPostgresProfile.scala

```diff
@@ -115,6 +117,89 @@ trait ExPostgresProfile extends JdbcProfile with PostgresProfile with Logging {
     override def transformMapping(n: Node) = reorderColumns(n, insertingSyms ++ nonPkAutoIncSyms ++ nonPkAutoIncSyms ++ nonPkAutoIncSyms)
   }
 
+  class MultiUpsertBuilder(ins: Insert) extends UpsertBuilderBase(ins) {
+    override def buildInsert: InsertBuilderResult = {
+      val start = allNames.iterator.mkString(s"insert into $tableName (", ",", ") ")
+      val insert = s"$start values $allVars"
+      val conflictWithPadding = "conflict (" + pkNames.mkString(", ") + ")" + (
+        if (nonPkAutoIncSyms.isEmpty) ""
+        else "where ? is null or ?=?"
+      )
+      val updateOrNothing =
+        if (softNames.isEmpty) "nothing"
+        else "update set " + softNames.map(n => s"$n=EXCLUDED.$n").mkString(",")
+      new InsertBuilderResult(table, s"$insert on $conflictWithPadding do $updateOrNothing", syms)
+    }
+
+    override def transformMapping(n: Node) =
+      reorderColumns(n, insertingSyms ++ nonPkAutoIncSyms ++ nonPkAutoIncSyms ++ nonPkAutoIncSyms)
+  }
+
+  class MultiUpsertCompiledInsert(node: Node) extends JdbcCompiledInsert(node) {
+    lazy val multiUpsert = compile(multiUpsertCompiler)
+  }
+
+  implicit def multiUpsertExtensionMethods[U, C[_]](q: Query[_, U, C]): InsertActionComposerImpl[U] =
+    new InsertActionComposerImpl[U](compileInsert(q.toNode))
+
+  lazy val multiUpsertCompiler = QueryCompiler(
+    Phase.assignUniqueSymbols,
+    Phase.inferTypes,
+    new InsertCompiler(InsertCompiler.AllColumns),
+    new JdbcInsertCodeGen(insert => new MultiUpsertBuilder(insert)))
```

Comment:
maybe you need change it to `new JdbcInsertCodeGen(insert => new NativeUpsertBuilder(insert))`

Reply:
I don't think so - I have a version of this code inside a private repo and it works, functionally. The part that seems to be failing is that in the test, it's not picking up the upsert capability, so it goes into the branch that throws instead of performing the operation.